### ISSUE TITLE
Flipout implementation for recurrent networks (lstm, gru) 

### ISSUE THREAD

**Issue by lbasora** [TyXe-BDL/TyXe#6]
Thanks for the excellent TyXe initiative. 
Currently, flipout is implemented in TyXe for linear and convolutional layers.  Are you considering supporting as well RNN in the near future?
If I understood well, for linear and conv layers you monkey-patched F.linear and F.conv, but I didn't see any equivalent functions for RNNs in torch.nn.functional allowing for a similar solution. Do you have any idea on how to implement this?

**Comment by lbasora** [comment]
To avoid re-implementing the forward pass, are you suggesting TyXe should provide the user with LSTM/GRU Flipout layers? As in [bayesian-torch](https://github.com/IntelLabs/bayesian-torch/blob/main/bayesian_torch/layers/flipout_layers/rnn_flipout.py) or in [Edward2](https://github.com/google/edward2/blob/main/edward2/tensorflow/layers/recurrent.py)? But then what about the nice TyXe feature of turning a regular pytorch LSTM/GRU model into a BNN model?

I agree with you current implementation isn't correct because as you say a new flipout mask is applied at each time step. But I don't know yet how we can cache the sign multipliers. Let me know if you see a practial way to do that.

Good suggestion for the testing.

**Comment by lbasora** [comment]
I've commited some changes (d77b8c69ecb8bb2e4e8768cbdb596f2a729c3897) to try to take into account your last comments. Please let me know whether the interface is convenient for you or you want some changes. 

I've tested and we obtain the same results that with the pytorch GRU/LSTM implementation when not using flipout. I also did the basic sanity check you suggested: when input is repeated along the batch dimension inside the flipout context we get different outputs whereas outside they are identical. I haven't tested though the consistency regarding the cached sign multipliers.

The current LSTM/GRU is not complete yet. Features like multilayer, bidirectionality and dropout can be added later though.
